File: solutions/erlang/beer-song/1/src/beer_song.erl
-module(beer_song).

-export([verse/1, sing/1, sing/2]).

-define(STARTING_COUNT, 99).
-define(PENULTIMATE_BOTTLE, 2).
-define(FINAL_BOTTLE, 1).
-define(NO_MORE_BOTTLES, 0).
-define(REPLACE_BOTTLES, -1).

verse(?NO_MORE_BOTTLES) ->
  BasicVerse = basicVerse(?NO_MORE_BOTTLES),
  FixedVerse = lists:flatten(string:replace(
    BasicVerse,
    "Take one down and pass it around,", "Go to the store and buy some more,"
  )),
  declareNoBottles(FixedVerse);

verse(?FINAL_BOTTLE) ->
  BasicVerse = basicVerse(?FINAL_BOTTLE),
  FixedVerse = lists:flatten(string:replace(BasicVerse, "Take one", "Take it")),
  SingularVerse = singularizeLastBottle(FixedVerse),
  declareNoBottles(SingularVerse);

verse(?PENULTIMATE_BOTTLE) ->
  BasicVerse = basicVerse(?PENULTIMATE_BOTTLE),
  singularizeLastBottle(BasicVerse);

verse(Integer) ->
  basicVerse(Integer).

basicVerse(Integer) ->
  integer_to_list(Integer) ++ " bottles of beer on the wall, " ++ integer_to_list(Integer) ++ " bottles of beer.\n" ++
    "Take one down and pass it around, " ++ calculateAfterPassing(Integer) ++ " bottles of beer on the wall.\n".

singularizeLastBottle(Verse) ->
  string:replace(Verse, "1 bottles", "1 bottle", all).

declareNoBottles(Verse) ->
  ReplacedVerse = string:replace(Verse, "0 bottles", "No more bottles", all),
  string:replace(ReplacedVerse, " No", " no").

calculateAfterPassing(?NO_MORE_BOTTLES) ->
  integer_to_list(?STARTING_COUNT);

calculateAfterPassing(Integer) ->
  integer_to_list(Integer - 1).

sing(From) ->
  sing(From, ?NO_MORE_BOTTLES).

sing(From, To) ->
  Range = lists:seq(From, To, -1),
  lists:foldl(fun addVerse/2, [], Range).

addVerse(Bottles, PastVerses) ->
  PastVerses ++ verse(Bottles) ++ "\n".

File: solutions/elixir/beer-song/1/lib/beer_song.ex
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @no_bottles "no more bottles"
  @one_bottle "1 bottle"

  @spec verse(integer) :: String.t()
  def verse(number),
    do:
      verse(
        _before_passing = calculate(number),
        _after_passing = calculate(number - 1)
      )

  defp verse(before_passing = @no_bottles, _after_passing),
    do:
      verse(
        before_passing,
        _after_passing = calculate(99),
        _directive = "Go to the store and buy some more, "
      )

  defp verse(before_passing, after_passing),
    do: verse(before_passing, after_passing, _directive = order_take(before_passing))

  defp verse(before_passing, after_passing, directive),
    do: "#{recite_count(before_passing)}#{directive}#{recite_remaining(after_passing)}"

  defp recite_count(before_passing),
    do: "#{proper_case(before_passing)} of beer on the wall, #{before_passing} of beer.\n"

  defp recite_remaining(after_passing),
    do: "#{after_passing} of beer on the wall.\n"

  defp order_take(@one_bottle),
    do: take("it")

  defp order_take(_),
    do: take("one")

  defp take(what),
    do: "Take #{what} down and pass it around, "

  defp proper_case(string),
    do: String.replace(string, "n", "N")

  defp calculate(0),
    do: @no_bottles

  defp calculate(1),
    do: @one_bottle

  defp calculate(number),
    do: "#{number} bottles"

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range \\ 99..0),
    do:
      range
      |> Enum.map(&verse/1)
      |> Enum.join("\n")
end
